File: analiseFreq.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class analiseFreq {

    // Dicionário de substituição de caracteres acentuados
    // Este mapa associa caracteres acentuados a seus equivalentes sem acento.
    // Isso é útil para normalizar o texto antes de qualquer análise ou decifração.
    private static final Map<Character, Character> substituicoes = new HashMap<>();
    
    static {
        // Mapeamento dos caracteres acentuados para seus equivalentes sem acento.
        substituicoes.put('á', 'a'); substituicoes.put('à', 'a'); substituicoes.put('ã', 'a');
        substituicoes.put('â', 'a'); substituicoes.put('ä', 'a'); substituicoes.put('é', 'e');
        substituicoes.put('è', 'e'); substituicoes.put('ê', 'e'); substituicoes.put('ë', 'e');
        substituicoes.put('í', 'i'); substituicoes.put('ì', 'i'); substituicoes.put('î', 'i');
        substituicoes.put('ï', 'i'); substituicoes.put('ó', 'o'); substituicoes.put('ò', 'o');
        substituicoes.put('õ', 'o'); substituicoes.put('ô', 'o'); substituicoes.put('ö', 'o');
        substituicoes.put('ú', 'u'); substituicoes.put('ù', 'u'); substituicoes.put('û', 'u');
        substituicoes.put('ü', 'u'); substituicoes.put('ç', 'c');
    }

    // Frequências típicas das letras na língua portuguesa
    // Este mapa armazena as frequências esperadas de cada letra em textos em português.
    // Esses valores são usados para calcular o valor de Chi-Quadrado e determinar o melhor deslocamento.
    private static final Map<Character, Double> frequenciasPortugues = new HashMap<>();
    
    static {
        // As frequências foram obtidas a partir de análises estatísticas da língua portuguesa.
        frequenciasPortugues.put('a', 14.63); frequenciasPortugues.put('b', 1.04);
        frequenciasPortugues.put('c', 3.88); frequenciasPortugues.put('d', 4.99);
        frequenciasPortugues.put('e', 12.57); frequenciasPortugues.put('f', 1.02);
        frequenciasPortugues.put('g', 1.30); frequenciasPortugues.put('h', 1.28);
        frequenciasPortugues.put('i', 6.18); frequenciasPortugues.put('j', 0.40);
        frequenciasPortugues.put('k', 0.02); frequenciasPortugues.put('l', 2.78);
        frequenciasPortugues.put('m', 4.74); frequenciasPortugues.put('n', 5.05);
        frequenciasPortugues.put('o', 10.73); frequenciasPortugues.put('p', 2.52);
        frequenciasPortugues.put('q', 1.20); frequenciasPortugues.put('r', 6.53);
        frequenciasPortugues.put('s', 7.81); frequenciasPortugues.put('t', 4.34);
        frequenciasPortugues.put('u', 4.63); frequenciasPortugues.put('v', 1.67);
        frequenciasPortugues.put('w', 0.01); frequenciasPortugues.put('x', 0.21);
        frequenciasPortugues.put('y', 0.01); frequenciasPortugues.put('z', 0.47);
    }

    /**
     * Normaliza o texto substituindo caracteres acentuados por suas versões sem acento.
     * A função também preserva a capitalização original dos caracteres.
     *
     * @param texto O texto a ser normalizado.
     * @return O texto normalizado.
     */
    public static String normalizarTexto(String texto) {
        StringBuilder textoNormalizado = new StringBuilder();

        // Percorre cada caractere do texto original.
        for (char c : texto.toCharArray()) {
            // Converte o caractere para minúsculo para facilitar a busca no mapa de substituições.
            char lowerC = Character.toLowerCase(c);
            
            // Verifica se o caractere está no mapa de substituições.
            if (substituicoes.containsKey(lowerC)) {
                // Se o caractere original era maiúsculo, converte a substituição para maiúsculo.
                if (Character.isUpperCase(c)) {
                    textoNormalizado.append(Character.toUpperCase(substituicoes.get(lowerC)));
                } else {
                    // Caso contrário, usa a versão minúscula do substituto.
                    textoNormalizado.append(substituicoes.get(lowerC));
                }
            } else {
                // Se o caractere não precisa ser substituído, simplesmente o adiciona ao texto normalizado.
                textoNormalizado.append(c);
            }
        }

        return textoNormalizado.toString();
    }

    /**
     * Calcula a frequência de cada letra no texto fornecido.
     * A frequência é expressa como uma porcentagem do total de letras no texto.
     *
     * @param texto O texto cuja frequência de letras será calculada.
     * @return Um mapa com as frequências das letras no texto.
     */
    public static Map<Character, Double> calcularFrequencias(String texto) {
        Map<Character, Double> frequencias = new HashMap<>();
        int totalLetras = 0;

        // Percorre cada caractere do texto normalizado.
        for (char c : texto.toLowerCase().toCharArray()) {
            // Verifica se o caractere é uma letra (ignorando espaços, pontuação, etc.).
            if (Character.isLetter(c)) {
                totalLetras++;
                // Incrementa a contagem da letra no mapa de frequências.
                frequencias.put(c, frequencias.getOrDefault(c, 0.0) + 1);
            }
        }

        // Converte as contagens absolutas de cada letra em porcentagens.
        for (char c : frequencias.keySet()) {
            frequencias.put(c, (frequencias.get(c) / totalLetras) * 100);
        }

        return frequencias;
    }

    /**
     * Calcula o valor do Chi-Quadrado para um determinado deslocamento no alfabeto.
     * Este valor mede a diferença entre as frequências de letras no texto e as frequências esperadas
     * na língua portuguesa. Quanto menor o valor, mais provável que o texto esteja decifrado corretamente.
     *
     * @param frequenciasTexto Mapa com as frequências das letras no texto cifrado.
     * @param deslocamento O deslocamento aplicado ao alfabeto para este cálculo.
     * @return O valor do Chi-Quadrado.
     */
    public static double calcularChiQuadrado(Map<Character, Double> frequenciasTexto, int deslocamento) {
        double chiQuadrado = 0.0;

        // Percorre cada entrada do mapa de frequências da língua portuguesa.
        for (Map.Entry<Character, Double> entry : frequenciasPortugues.entrySet()) {
            // Calcula o caractere após aplicar o deslocamento no alfabeto.
            char charDeslocado = (char) (((entry.getKey() - 'a' + deslocamento) % 26) + 'a');
            // Obtém a frequência do caractere deslocado no texto, ou assume 0 se não estiver presente.
            double freqTexto = frequenciasTexto.getOrDefault(charDeslocado, 0.0);

            // Aplica a fórmula do Chi-Quadrado: (observado - esperado)² / esperado.
            chiQuadrado += Math.pow(freqTexto - entry.getValue(), 2) / entry.getValue();
        }

        return chiQuadrado;
    }

    /**
     * Decifra um texto cifrado usando a cifra de César, identificando o melhor deslocamento
     * através da minimização do Chi-Quadrado.
     *
     * @param textoCifrado O texto cifrado que será decifrado.
     * @return Um array de strings, onde o primeiro elemento é o texto decifrado e o segundo é o deslocamento usado.
     */
    public static String[] decifrarCesar(String textoCifrado) {
        // Normaliza o texto cifrado para remover acentos e unificar o formato das letras.
        String textoNormalizado = normalizarTexto(textoCifrado);
        // Calcula as frequências de letras no texto cifrado normalizado.
        Map<Character, Double> frequenciasTexto = calcularFrequencias(textoNormalizado);

        int melhorDeslocamento = 0;
        double menorChiQuadrado = Double.POSITIVE_INFINITY;

        // Testa todos os deslocamentos possíveis (de 0 a 25).
        for (int deslocamento = 0; deslocamento < 26; deslocamento++) {
            // Calcula o valor do Chi-Quadrado para o deslocamento atual.
            double chiQuadrado = calcularChiQuadrado(frequenciasTexto, deslocamento);
            // Se o valor do Chi-Quadrado for menor que o menor valor encontrado até agora,
            // atualiza o melhor deslocamento.
            if (chiQuadrado < menorChiQuadrado) {
                menorChiQuadrado = chiQuadrado;
                melhorDeslocamento = deslocamento;
            }
        }

        StringBuilder textoDecifrado = new StringBuilder();

        // Decifra o texto usando o deslocamento identificado.
        for (char c : textoNormalizado.toCharArray()) {
            if (Character.isLetter(c)) {
                // Calcula o caractere decifrado aplicando o deslocamento inverso.
                if (Character.isLowerCase(c)) {
                    textoDecifrado.append((char) (((c - 'a' - melhorDeslocamento + 26) % 26) + 'a'));
                } else {
                    textoDecifrado.append((char) (((c - 'A' - melhorDeslocamento + 26) % 26) + 'A'));
                }
            } else {
                // Se o caractere não é uma letra, adiciona-o diretamente ao texto decifrado.
                textoDecifrado.append(c);
            }
        }

        return new String[]{textoDecifrado.toString(), String.valueOf(melhorDeslocamento)};
    }

    public static void main(String[] args) {
        // Recebe o texto cifrado como entrada do usuário.
        Scanner scanner = new Scanner(System.in);
        String textoCifrado = scanner.nextLine();

        // Decifra o texto e obtém o deslocamento utilizado na cifra.
        String[] resultado = decifrarCesar(textoCifrado);

        // Exibe o deslocamento encontrado e o texto decifrado.
        System.out.println(resultado[1]);
        System.out.println(resultado[0]);
    }
}

File: analiseFreq.py
# Dicionário de substituição de caracteres acentuados
# Este dicionário mapeia caracteres acentuados para seus equivalentes sem acento,
# o que ajuda a normalizar o texto para análise posterior.
substituicoes = {
    'á': 'a', 'à': 'a', 'ã': 'a', 'â': 'a', 'ä': 'a', 'é': 'e', 'è': 'e', 'ê': 'e', 'ë': 'e',
    'í': 'i', 'ì': 'i', 'î': 'i', 'ï': 'i', 'ó': 'o', 'ò': 'o', 'õ': 'o', 'ô': 'o', 'ö': 'o',
    'ú': 'u', 'ù': 'u', 'û': 'u', 'ü': 'u', 'ç': 'c'
}

def normalizar_texto(texto):
    """
    Normaliza o texto substituindo caracteres acentuados por suas versões sem acento.
    Também preserva a capitalização original dos caracteres.
    """
    texto_normalizado = ""
    for char in texto:
        # Verifica se o caractere é acentuado, e substitui conforme o dicionário
        if char.lower() in substituicoes:
            if char.isupper():
                texto_normalizado += substituicoes[char.lower()].upper()
            else:
                texto_normalizado += substituicoes[char.lower()]
        else:
            texto_normalizado += char
    return texto_normalizado

# Frequências típicas das letras na língua portuguesa
# Estes valores são usados como referência para calcular o Chi-Quadrado
frequencias_portugues = {
    'a': 14.63,
    'b': 1.04,
    'c': 3.88,
    'd': 4.99,
    'e': 12.57,
    'f': 1.02,
    'g': 1.30,
    'h': 1.28,
    'i': 6.18,
    'j': 0.40,
    'k': 0.02,
    'l': 2.78,
    'm': 4.74,
    'n': 5.05,
    'o': 10.73,
    'p': 2.52,
    'q': 1.20,
    'r': 6.53,
    's': 7.81,
    't': 4.34,
    'u': 4.63,
    'v': 1.67,
    'w': 0.01,
    'x': 0.21,
    'y': 0.01,
    'z': 0.47
}

def calcular_frequencias(texto):
    """
    Calcula a frequência de cada letra no texto.
    O resultado é expresso em porcentagem do total de letras no texto.
    """
    frequencias = {}
    total_letras = 0
    for char in texto.lower():
        if char.isalpha():
            total_letras += 1
            if char in frequencias:
                frequencias[char] += 1
            else:
                frequencias[char] = 1
    # Converte as contagens absolutas em porcentagens
    for char in frequencias:
        frequencias[char] = (frequencias[char] / total_letras) * 100
    return frequencias

def calcular_chi_quadrado(frequencias_texto, deslocamento):
    """
    Calcula o valor do Chi-Quadrado para um determinado deslocamento no alfabeto.
    O Chi-Quadrado mede a diferença entre as frequências de letras no texto e 
    as frequências esperadas na língua portuguesa, considerando um deslocamento
    aplicado ao texto cifrado.
    """
    chi_quadrado = 0.0
    for char, freq in frequencias_portugues.items():
        # Calcula o caractere após aplicar o deslocamento no alfabeto
        char_deslocado = chr(((ord(char) - ord('a') + deslocamento) % 26) + ord('a'))
        # Obtém a frequência do caractere deslocado no texto, ou assume 0 se não estiver presente
        freq_texto = frequencias_texto.get(char_deslocado, 0)
        # Soma a contribuição do caractere ao valor total de Chi-Quadrado
        # A fórmula é: (observado - esperado)² / esperado
        chi_quadrado += ((freq_texto - freq) ** 2) / freq
    return chi_quadrado

def decifrar_cesar(texto_cifrado):
    """
    Decifra um texto cifrado usando a cifra de César, identificando o melhor deslocamento
    através da minimização do Chi-Quadrado.
    """
    # Normaliza o texto cifrado para remover acentos e unificar o formato das letras
    texto_normalizado = normalizar_texto(texto_cifrado)
    
    # Calcula as frequências de letras no texto cifrado normalizado
    frequencias_texto = calcular_frequencias(texto_normalizado)

    # Inicializa variáveis para encontrar o melhor deslocamento
    melhor_deslocamento = 0
    menor_chi_quadrado = float('inf')
    
    # Testa todos os deslocamentos possíveis (0 a 25)
    for deslocamento in range(26):
        chi_quadrado = calcular_chi_quadrado(frequencias_texto, deslocamento)
        # Atualiza o melhor deslocamento se o valor do Chi-Quadrado for menor
        if chi_quadrado < menor_chi_quadrado:
            menor_chi_quadrado = chi_quadrado
            melhor_deslocamento = deslocamento

    # Decifra o texto usando o deslocamento identificado
    texto_decifrado = ''
    for char in texto_normalizado:
        if char.isalpha():
            # Calcula o caractere decifrado aplicando o deslocamento inverso
            if char.islower():
                texto_decifrado += chr(((ord(char) - ord('a') - melhor_deslocamento) % 26) + ord('a'))
            else:
                texto_decifrado += chr(((ord(char) - ord('A') - melhor_deslocamento) % 26) + ord('A'))
        else:
            texto_decifrado += char

    return texto_decifrado, melhor_deslocamento

# Recebe o texto cifrado como entrada do usuário
texto_cifrado = input()
# Decifra o texto e obtém o deslocamento utilizado na cifra
texto_decifrado, deslocamento = decifrar_cesar(texto_cifrado)

# Exibe o deslocamento encontrado e o texto decifrado
print(deslocamento)
print(texto_decifrado)
